natter-api/src/main/java/com/manning/apisecurityinaction/controller/SpaceController.java

```diff
@@ -50,85 +56,120 @@ public JSONObject createSpace(Request request, Response response) {
   public JSONObject postMessage(Request request, Response response) {
     var spaceId = Long.parseLong(request.params(":spaceId"));
     var json = new JSONObject(request.body());
+
     var user = json.getString("author");
+
     if (!user.matches("[a-zA-Z][a-zA-Z0-9]{0,29}")) {
+
       throw new IllegalArgumentException("invalid username");
+
     }
     var message = json.getString("message");
+
     if (message.length() > 1024) {
+
       throw new IllegalArgumentException("message is too long");
+
     }
 
+
     return database.withTransaction(tx -> {
+
       var msgId = database.findUniqueLong(
           "SELECT NEXT VALUE FOR msg_id_seq;");
+
       database.updateUnique(
+
           "INSERT INTO messages(space_id, msg_id, msg_time," +
               "author, msg_text) " +
               "VALUES(?, ?, current_timestamp, ?, ?)",
+
           spaceId, msgId, user, message);
 
       response.status(201);
+
       var uri = "/spaces/" + spaceId + "/messages/" + msgId;
+
       response.header("Location", uri);
+
       return new JSONObject().put("uri", uri);
+
     });
   }
 
   public Message readMessage(Request request, Response response) {
     var spaceId = Long.parseLong(request.params(":spaceId"));
-    var msgId = Long.parseLong(request.params(":msgId"));
 
+    var msgId = Long.parseLong(request.params(":msgId"));
     var message = database.findUnique(Message.class,
+
         "SELECT space_id, msg_id, author, msg_time, msg_text " +
             "FROM messages WHERE msg_id = ? AND space_id = ?",
         msgId, spaceId);
-
     response.status(200);
     return message;
   }
 
   public JSONArray findMessages(Request request, Response response) {
```

Comment:
Code review for findMessages function:

1. spaceId should not be added in url and it can be added if it is encrypted
2. There is no check that the spaceId is owned by logged in user, so hacker can search message from any space
3. No error handling for missing input/param
4. No error handling for forbidden status
5. Description of function and parameters is not added
6. Case no results found does not exist